VNR-3895, REDCORE-503 #comment UTC filter for date/time and show time in datepicker(option)

shandak opened the pull request.

jatitoam commented:
shortcut @Kixo :)
I'd really like your review here

phproberto commented:
I really don't understand why don't you create a new field for the date time When you start mixing things is when things become too complex to maintain them. If you need a date time picker then create a new field. Cleaner and safer because you will be sure you aren't breaking things.

Also why are you changing tables, models, etc.? Isn't this about a form field? 1 PR -> 1 functionality/bug. Easy to review.

phproberto commented on extensions/libraries/redcore/form/fields/rdatepicker.php:
```diff
@@ -49,6 +51,109 @@ class JFormFieldRdatepicker extends JFormFieldRtext
+	 */
+	public function setup(SimpleXMLElement $element, $value, $group = null)
+	{
+		$return = parent::setup($element, $value, $group);
```
Why don't you add `showTime` attribute here too?

shandak replied:
Yeah, I'll fix it.

shandak replied:
done

phproberto commented on extensions/libraries/redcore/form/fields/rdatepicker.php:
```diff
@@ -61,6 +166,54 @@ protected function getInput()
+
+		if (isset($this->element['showTime']) && $this->element['showTime'] == 'true')
+		{
+			$phpFormat .= ' H:i:s';
```
This is where you start to see that mixing two concepts is a wrong idea. I wouldn't add a hardcoded time format here. Also you already have a function to retrieve the format (`dateFormatJQueryUIToPHP`). Why process things again outside it?

phproberto commented on extensions/libraries/redcore/form/fields/rdatepicker.php:
```diff
@@ -77,13 +230,95 @@ protected function getInput()
 	protected function getDatepickerOptions()
 	{
-		$options = new JRegistry;
+		$return = '{';
```
Object: Easy to debug and convert to other formats
String: Hardcoded shit easy to break

shandak replied:
Agree, I changed it, because it converted JSON string values always with quotes, but sometimes we can send anonymous function or regular function like value and quotes, in this case, don't need. This is the regular situation for javascript - use JSON construction, where value can be a function. But PHP json_encode can not allow it.

shandak replied:
I didn't get you. What don't you see? There is https://github.com/redCOMPONENT-COM/redCORE/pull/744/files#diff-539a740304d8b474cac4672d6cb24bb5R395 an example where we can set function instead of text.

shandak commented:
The main issue which I have done here - it adds UTC conversion for datepicker (wich use optional for do not break all already developed extensions). After that, we have possibility safe change it in all related extensions. If you do not have any special field elements in the xml, then all must be work as before changes.
When I tried fixed it - I found an extra code in the model(the same executed again in JTable), so I removed it.
After I changed the logic for function updateAuditFields https://github.com/redCOMPONENT-COM/redCORE/blob/develop/extensions/libraries/redcore/table/table.php#L988 for use UTC conversion, but here was 80% duplicated code for RTable and RTableNested , so I created a trait for it. So it will support easily in future.
Of corse, I could create a new field, but I just expand several new possibilities for field, but not created something new.

phproberto commented:
You are:
- Adding traits which requires php 5.4. AFAIK that changes redCORE requirements & needs that you are sure what are you doing because it will affect all the sites. I'm ok raising requirements but that probably requires a new major version.
- Changing format of the value. So any override expecting value in X format may get broken.
- Changing tables that may break other things and cause a WTF when people sees it was caused by a PR with a title like `UTC filter for date/time`

shandak commented:
Yeah, you are right about php5.4 and traits, I missed it. What do you think about it @jatitoam @Kixo would I bump redcore version and  change PHP requirement to php5.4 or just remove trait?
About second and third - I didn't get you. It shouldn't cause any problem if user didn't setup filter="utc_(user/server)" and/or showTime="true". It must work the same like before changes, nothing logic changes without new parameters.

Kixo commented:
I am sorry but we should follow at least latest stable Joomla requirements, so it is PHP 5.3.10
We still have many extensions that are working on old sites and depending on redCORE, raising the bar to PHP 5.4 could prove way more problematic for them.
I agree that Traits are great and wanted to use them myself but we are restricted to the PHP versioning, so not yet...

shandak commented:
I've removed trait

shandak commented on extensions/libraries/redcore/model/list.php:
```diff
@@ -454,6 +454,9 @@ protected function loadForm($name, $source = null, $options = array(), $clear =
 			$this->preprocessForm($form, $data);
 
+			// Filter and validate the form data.
+			$data = $form->filter($data);
```
@Kixo @phproberto @jatitoam please take a look at this string. I'd prefer to use data filter for filter forms also. It doesn't use in joomla core. This is data filter will help to convert datetime fields in the right UTC offset.
What do you think about it?

shandak replied:
It is not

Kixo commented on extensions/libraries/redcore/model/admin/legacy.php:
```diff
@@ -9,96 +9,13 @@
 
 defined('JPATH_REDCORE') or die;
 
-if (version_compare(JVERSION, '3.0', 'lt'))
```
@shandak bump

Kixo replied:
yes but it does not populate those fields in joomla < 3

shandak replied:
Fields will populate `JTable` `updateAuditFields()`.
I don't see any reason why can be skip `updateAuditFields()` in joomla < 3.
Did you checked it in the real site or it is just your assumption?

shandak commented on extensions/libraries/redcore/form/fields/rdatepicker.php:
```diff
@@ -49,6 +49,109 @@ class JFormFieldRdatepicker extends JFormFieldRtext
+	 *
+	 * @return  mixed  The property value or null.
+	 */
+	public function __get($name)
```
Because there is a standard joomla approach. I just expanded already existed magic method in the joomla class.
